### PR TITLE
Feat/sp2 1794 cookie management

### PR DESCRIPTION
**Pull request by CoburnJoe**
First pass at cookie management:
- Pipelines config
- Cookie manager + tests

Still todo: typehint the default logger type of lambdas. Lock down dependencies, docs, implementation guide, packaging etc.

### CODE REVIEW

File: cookie_manager/cookie_manager.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+    def _ratify_config(self, override_config: dict) -> dict:
+        """
+        Takes a config dict ``override_config`` and overrides any default values found in ``self._config``
+        Will error out if a config element does not exist in the parent dict
+        :param override_config: Dict of config options
+        :return: copy of ``self._config`` dict with values updated by those supplied in ``override_config``
+        """
+        self._logger.info(f"Starting to compare config: {override_config}")
+        new_config: dict = self._config
```
Comment (Bekabyx) [inline cookie_manager/cookie_manager.py]: How come this is the only line where we're doing typing?

File: cookie_manager/cookie_manager.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+        {
+            "error": lambda msg: print(msg),
+            "warning": lambda msg: print(msg),
+            "debug": lambda msg: print(msg),
+            "info": lambda msg: print(msg),
+        },
+    )
+
+    def __init__(self, config: dict = None, logger=None) -> None:
+        self._logger.info(f"Configuring cookie manager")
```
Comment (nomad-software) [inline cookie_manager/cookie_manager.py]: The logger should be set first before any logging takes place.

File: cookie_manager/cookie_manager.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        if config:
+            self._config = self._ratify_config(override_config=config)
+        if logger:
+            self._logger = logger
+        self._logger.info(f"Finished configuring cookie manager")
+
+    def _ratify_config(self, override_config: dict) -> dict:
```
Comment (nomad-software) [inline cookie_manager/cookie_manager.py]: `_override_config` might be a better name?

File: cookie_manager/cookie_manager.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        """
+        Takes a config dict ``override_config`` and overrides any default values found in ``self._config``
+        Will error out if a config element does not exist in the parent dict
+        :param override_config: Dict of config options
+        :return: copy of ``self._config`` dict with values updated by those supplied in ``override_config``
+        """
+        self._logger.info(f"Starting to compare config: {override_config}")
+        new_config: dict = self._config
+
+        if override_config:
```
Comment (nomad-software) [inline cookie_manager/cookie_manager.py]: This check can be removed as it is checked in the constructor.

File: cookie_manager/cookie_manager.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
+import json
+import re
+
+from itsdangerous import TimestampSigner, BadSignature, SignatureExpired
+from werkzeug.exceptions import Unauthorized, ServiceUnavailable, BadRequest
```
Comment (nomad-software) [inline cookie_manager/cookie_manager.py]: Should this library be responsible for raising request errors or should that be a job for middleware?

File: cookie_manager/cookie_manager.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+        :return: Verified cookie dict payload, None,
+        """
+        self._logger.info(f"Starting to verify cookie: {cookie}")
+
+        if cookie is None:
+            self._logger.warning(f"Incoming cookie '{cookie}' not provided.")
+            raise Unauthorized
+        try:
+            self._logger.debug(f"Beginning verification: {cookie}")
+            cookie_value_json = TimestampSigner(verify_key).unsign(
```
Comment (nomad-software) [inline cookie_manager/cookie_manager.py]: Here we're assuming the cookie value is Json but it that may not necessarily be so.

File: cookie_manager/cookie_manager.py (hunk 5 above)
Comment (CoburnJoe) [reply cookie_manager/cookie_manager.py]: Possibly not - I'd like to suggest redesigning this bit on the next pass.

File: cookie_manager/cookie_manager.py (hunk 2 above)
Comment (CoburnJoe) [reply cookie_manager/cookie_manager.py]: Yes good catch

File: cookie_manager/cookie_manager.py (hunk 3 above)
Comment (CoburnJoe) [reply cookie_manager/cookie_manager.py]: Done

File: cookie_manager/cookie_manager.py (hunk 6 above)
Comment (CoburnJoe) [reply cookie_manager/cookie_manager.py]: Done

File: cookie_manager/cookie_manager.py (hunk 4 above)
Comment (CoburnJoe) [reply cookie_manager/cookie_manager.py]: I disagree -- this check exists to defensively handle values of `None` or other None types such as `[]`, `{}`, or `""`. Tested in `test_override_config_positive` which raised the bug (if this line isn't present).

Comment (nomad-software) [reply cookie_manager/cookie_manager.py]: ok

**Comment by CoburnJoe** [comment]
> It appears that we use static typing 2 or 3 times sporadically and that we don't have any libraries for type checking. Is this deliberate?

Yes - as is the plan with many of our SP projects, we want to start using static typing everywhere. Static typing was deliberately addd to this codebase, with the exception of one or two variables which are defined using `type` -> we can annotate these with `Callable([str, str], str)`, but it's still not quite the right annotation, so still thinking about how best to solve this bit.

**Pull request merged**